File: classes/game.js
const tetrominoes = [T, Straight, Square, J, L, Z, S];

class Game {
  constructor(height) {
    let board = [];
    for (let i = 0; i < 22; i++) {
      board.push([]);
      for (let j = 0; j < 10; j++) {
        board[i].push(null);
      }
    }
    this.setup = board;
    this.height = height;
    let rows = [];
    rows.length = 22;
    rows.fill(0);
    this.rows = rows;
    this.dropping = this.getRandomTetromino();
    this.next = this.getRandomTetromino();
    this.holding = null;
    this.switchAllowed = true;
  }

  getRandomTetromino() {
    let index = floor(random(7));
    return new tetrominoes[index](colors[index], this.height / 22, this.setup);
  }

  up() {
    while (true) {
      if (this.drop() === 0) break;
    }
  }

  right() {
    for (let square of this.dropping.squares) {
      if (square.x + 1 > 9 || this.setup[square.y][square.x + 1]) return;
    }
    for (let square of this.dropping.squares) square.x += 1;
  }

  down() {
    this.drop();
  }

  left() {
    for (let square of this.dropping.squares) {
      if (square.x - 1 < 0 || this.setup[square.y][square.x - 1]) return;
    }
    for (let square of this.dropping.squares) square.x -= 1;
  }

  switch() {
    if (this.switchAllowed) {
      this.switchAllowed = false;
      console.log(this.dropping.squares[0].img);
      console.log(this.dropping.reset);
      this.dropping.reset();
      if (this.holding) {
        let temp = this.holding;
        this.holding = this.dropping;
        this.dropping = temp;
        return;
      }
      console.log(this.dropping.squares[0].img);
      this.holding = this.dropping;
      console.log(this.holding.squares[0].img);
      this.dropping = this.getRandomTetromino();
      console.log("a");
    }
  }

  draw(height, squareSize) {
    background(BACKGROUND);
    for (let row of this.setup) {
      for (let square of row) {
        if (square) square.draw(squareSize, 0, undefined);
      }
    }

    this.dropping.draw(squareSize);
    let x = squareSize * 6;
    // left line of the board
    stroke(LINES);
    line(x, 0, x, height);
    // HOLD section
    fill(BACKGROUND);
    rect(0, 1, x, x);
    noStroke();
    fill(LINES);
    text("HOLD:", 1, 14);
    if (this.holding) this.holding.draw(squareSize, 1);

    stroke(LINES);
    line(x, height, x + squareSize * 10, height);
    // right line of the board
    x = squareSize * 16;
    line(x, 1, x, height);
    // NEXT section
    noFill();
    rect(x, 0, height - x, height - x);
    noStroke();
    fill(LINES);
    text("NEXT:", x + 1, 14);
    this.next.draw(squareSize, 2);
  }

  drop() {
    for (let block of this.dropping.squares) {
      if (block.y >= 21 || this.setup[block.y + 1][block.x]) {
        this.startDrop();
        return 0;
      }
    }
    this.dropping.drop();
  }

  startDrop() {
    this.addToRows();
    for (let block of this.dropping.squares)
      this.setup[block.y][block.x] = block;
    this.dropping = this.next;
    this.next = this.getRandomTetromino();
    this.fullRow();
    this.switchAllowed = true;
  }

  addToRows() {
    for (let square of this.dropping.squares) this.rows[square.y]++;
  }

  fullRow() {
    for (let row = 0; row < 22; row++) {
      if (this.rows[row] === 10) {
        this.rows[row] = 0;
        this.setup.splice(row, 1);
        let insert = [];
        insert.length = 10;
        insert.fill(null);
        this.setup.unshift(insert);
        this.rows.splice(row, 1);
        this.rows.unshift(0);
        for (let i = 0; i <= row; i++) {
          for (let j = 0; j < 10; j++) {
            if (this.setup[i][j]) this.setup[i][j].y++;
          }
        }
      }
    }
  }
  static lose() {
    console.log("you lost");
  }
}

File: classes/game.py
"""The game itself happens here"""
import secrets

import numpy as np
import pygame

from classes.tetrominoes import Square, Straight, S, Z, L, J, T
from colors.colors import LINES, BACKGROUND

tetrominoes = [T, Straight, Square, J, L, Z, S]


class Game:
    """This is the class that represents everything that is happening"""
    def __init__(self, height):
        self.setup = np.empty((22, 10), dtype=object)
        self.height = height
        self.rows = np.zeros(22, dtype=np.uint8)
        self.dropping = self.get_random_tetromino()
        self.next = self.get_random_tetromino()
        self.holding = None
        self.switch_allowed = True

    def draw(self, window, height, square_size, font):
        """The entire window will be drawn here"""
        window.fill(BACKGROUND)
        # squares on the board
        for row in self.setup:
            for square in row:
                if square:
                    square.draw(window, square_size, 0)

        self.dropping.draw(window, square_size)
        x = square_size * 6
        # left line of the board
        pygame.draw.line(window, LINES, (x, 0), (x, height))
        # HOLD section
        pygame.draw.rect(window, LINES, (0, 0, x + 1, x), 1)
        window.blit(font.render('Hold:', True, LINES), (1, 0))
        if self.holding:
            self.holding.draw(window, square_size, 1)
        # right line of the board
        x = square_size * 16
        pygame.draw.line(window, LINES, (x, 0), (x, height))
        # NEXT section
        pygame.draw.rect(window, LINES, (x, 0, height - x, height - x), 1)
        window.blit(font.render('Next:', True, LINES), (x + 1, 0))
        self.next.draw(window, square_size, 2)

        pygame.display.update()

    def start_drop(self):
        """When there is no tetromino, this function will create the next tetromino"""
        self.add_to_rows()
        for block in self.dropping.squares:
            self.setup[block.y][block.x] = block
        self.dropping = self.next
        self.next = self.get_random_tetromino()
        self.full_row()
        self.switch_allowed = True

    def drop(self):
        """Every second, the falling tetromino approaches the ground"""
        for block in self.dropping.squares:
            if block.y >= 21 or self.setup[block.y + 1][block.x]:
                self.start_drop()
                return 0
        self.dropping.drop()

    def left(self):
        for square in self.dropping.squares:
            if square.x - 1 < 0 or self.setup[square.y][square.x - 1]:
                return
        for square in self.dropping.squares:
            square.x -= 1

    def right(self):
        for square in self.dropping.squares:
            if square.x + 1 > 9 or self.setup[square.y][square.x + 1]:
                return
        for square in self.dropping.squares:
            square.x += 1

    def down(self):
        self.drop()

    def up(self):
        while True:
            if self.drop() == 0:
                break

    def full_row(self):
        for row in range(22):
            if self.rows[row] == 10:
                self.rows[row] = 0
                self.setup = np.append([np.empty(10, dtype=object)], np.delete(self.setup, row, axis=0), axis=0)
                self.rows = np.append(0, np.delete(self.rows, row))
                for i in range(row + 1):
                    for j in range(10):
                        if self.setup[i][j]:
                            self.setup[i][j].y += 1

    def add_to_rows(self):
        for square in self.dropping.squares:
            self.rows[square.y] += 1

    def switch(self):
        if self.switch_allowed:
            self.switch_allowed = False
            self.dropping.reset()
            if self.holding:
                self.dropping, self.holding = self.holding, self.dropping
                return
            self.holding = self.dropping
            self.dropping = self.next
            self.next = self.get_random_tetromino()

    def get_random_tetromino(self):
        return secrets.choice(tetrominoes)(self.height // 22, self.setup)
